pkg/cmd/util/print/print.go

```diff
@@ -5,28 +5,147 @@ import (
+func convertItemsForDisplay(scheme *runtime.Scheme, objs []runtime.Object, preferredVersions ...schema.GroupVersion) ([]runtime.Object, error) {
+	ret := []runtime.Object{}
+
+	for i := range objs {
+		obj := objs[i]
+		kinds, _, err := scheme.ObjectKinds(obj)
+		if err != nil {
+			return nil, err
+		}
+
+		// Gather all groups where the object kind is known.
+		groups := []string{}
+		for _, kind := range kinds {
+			groups = append(groups, kind.Group)
+		}
+
+		// if no preferred versions given, pass all group versions found.
+		if len(preferredVersions) == 0 {
+			defaultGroupVersions := []runtime.GroupVersioner{}
+			for _, group := range groups {
+				defaultGroupVersions = append(defaultGroupVersions, schema.GroupVersions(scheme.PrioritizedVersionsForGroup(group)))
+			}
+
+			defaultGroupVersioners := runtime.GroupVersioners(defaultGroupVersions)
+			convertedObject, err := scheme.ConvertToVersion(obj, defaultGroupVersioners)
+			if err != nil {
+				return nil, err
+			}
+			ret = append(ret, convertedObject)
+			continue
+		}
+
+		actualOutputVersion := schema.GroupVersion{}
+		// Find the first preferred version that contains the object kind group.
+		// If there are more groups for the given resource, prefer those that are first in the
+		// list of preferred versions.
+		for _, version := range preferredVersions {
+			for _, group := range groups {
+				if version.Group == group {
+					for _, externalVersion := range scheme.PrioritizedVersionsForGroup(group) {
+						if version == externalVersion {
+							actualOutputVersion = externalVersion
+							break
+						}
+						if actualOutputVersion.Empty() {
+							actualOutputVersion = externalVersion
+						}
+					}
+				}
+				if !actualOutputVersion.Empty() {
+					break
+				}
+			}
+			if !actualOutputVersion.Empty() {
+				break
+			}
+		}
+
+		// if no preferred version found in the list of given GroupVersions,
+		// attempt to convert to first GroupVersion that satisfies a preferred version
+		if len(actualOutputVersion.Version) == 0 {
+			preferredVersioners := []runtime.GroupVersioner{}
+			for _, gv := range preferredVersions {
+				preferredVersions = append(preferredVersions, gv)
+			}
+			preferredVersioner := runtime.GroupVersioners(preferredVersioners)
+			convertedObject, err := scheme.ConvertToVersion(obj, preferredVersioner)
+			if err != nil {
+				return nil, err
+			}
+
+			ret = append(ret, convertedObject)
+			continue
+		}
+
+		convertedObject, err := scheme.ConvertToVersion(obj, actualOutputVersion)
+		if err != nil {
+			return nil, err
+		}
+
+		ret = append(ret, convertedObject)
+	}
+
+	return ret, nil
+}
+
+// convertItemsForDisplayFromDefaultCommand returns a new list that contains parallel elements that have been converted to the most preferred external version
+// TODO: move this function into the core factory PrintObjects method
+// TODO: print-objects should have preferred output versions
```

Comment:
@juanvallejo mind creating an issue upstream so we don't forget about it?

Reply:
https://github.com/kubernetes/kubernetes/issues/65490